services/src/machine_learning/xgb_config_builder.rs

```diff
@@ -0,0 +1,1057 @@
+    }
+
+    pub fn custom_evaluation_fn(&self) -> &Option<CustomEvaluation> {
+        &self.custom_evaluation_fn
+    }
+
+    pub fn set_custom_evaluation_fn(&mut self, custom_evaluation_fn: Option<CustomEvaluation>) {
+        self.custom_evaluation_fn = custom_evaluation_fn;
+    }
+}
+
+enum Inclusion {
+    Open,
+    Closed,
+}
+
+struct Interval<T> {
+    min: T,
+    min_inclusion: Inclusion,
+    max: T,
+    max_inclusion: Inclusion,
+}
+
+impl<T: Display> Display for Interval<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let lower = match self.min_inclusion {
+            Inclusion::Closed => '[',
+            Inclusion::Open => '(',
+        };
+        let upper = match self.max_inclusion {
+            Inclusion::Closed => ']',
+            Inclusion::Open => ')',
+        };
+        write!(f, "{}{}, {}{}", lower, self.min, self.max, upper)
+    }
+}
+
+impl<T: PartialOrd + Display> Interval<T> {
+    fn new(min: T, min_inclusion: Inclusion, max: T, max_inclusion: Inclusion) -> Self {
+        Interval {
+            min,
+            min_inclusion,
+            max,
+            max_inclusion,
+        }
+    }
+
+    fn new_open_open(min: T, max: T) -> Self {
+        Interval::new(min, Inclusion::Open, max, Inclusion::Open)
+    }
+
+    fn new_open_closed(min: T, max: T) -> Self {
+        Interval::new(min, Inclusion::Open, max, Inclusion::Closed)
+    }
+
+    fn new_closed_closed(min: T, max: T) -> Self {
+        Interval::new(min, Inclusion::Closed, max, Inclusion::Closed)
+    }
+
+    fn contains(&self, val: &T) -> bool {
+        let min = match self.min_inclusion {
+            Inclusion::Closed => {
+                matches!(
+                    val.partial_cmp(&self.min),
+                    Some(Ordering::Equal | Ordering::Greater)
+                )
+            }
+            Inclusion::Open => {
+                matches!(val.partial_cmp(&self.min), Some(Ordering::Greater))
+            }
+        };
+
+        let max = match self.max_inclusion {
+            Inclusion::Closed => {
+                matches!(
+                    val.partial_cmp(&self.max),
+                    Some(Ordering::Equal | Ordering::Less)
+                )
+            }
+            Inclusion::Open => matches!(val.partial_cmp(&self.max), Some(Ordering::Less)),
+        };
+
+        min & max
+    }
+
+    fn validate(&self, val: &Option<T>, name: &str) -> Result<(), String> {
+        match val {
+            Some(ref val) => {
+                if self.contains(val) {
+                    Ok(())
+                } else {
+                    Err(format!(
+                        "Invalid value for '{}' parameter, {} is not in range {}.",
+                        name, &val, self
+                    ))
+                }
+            }
+            None => Ok(()),
+        }
+    }
+}
+
+/// Type of booster to use when training a [Booster](../struct.Booster.html) model.
```

Comment:
Where does this link to?

Reply:
That should not have been there.

07c52602

---

Commit MSG:

chore: remove comment